gonvme_fc.go

```diff
@@ -0,0 +1,359 @@
+	return &nvme
+}
+
+func (nvme *NVMeFC) getChrootDirectory() string {
+	s := nvme.options[ChrootDirectory]
+	if s == "" {
+		s = "/"
+	}
+	return s
+}
+
+func (nvme *NVMeFC) buildNVMeCommand(cmd []string) []string {
+	if nvme.getChrootDirectory() == "/" {
+		return cmd
+	}
+	command := []string{"chroot", nvme.getChrootDirectory()}
+	command = append(command, cmd...)
+	return command
+}
+
+func (nvme *NVMeFC) getFCHostInfo() ([]FCHBAInfo, error) {
+
+	match, err := filepath.Glob("/sys/class/fc_host/host*")
+	if err != nil {
+		fmt.Printf("Error gathering fc hosts: %v", err)
+		return []FCHBAInfo{}, err
+	}
+
+	var FCHostsInfo []FCHBAInfo
+	for _, m := range match {
+
+		var FCHostInfo FCHBAInfo
+		data, err := ioutil.ReadFile(path.Join(m, "port_name"))
+		if err != nil {
+			fmt.Printf("match: %s failed to read port_name file: %s", match, err.Error())
+			continue
+		}
+		FCHostInfo.PortName = strings.TrimSpace(string(data))
+
+		data, err = ioutil.ReadFile(path.Join(m, "node_name"))
+		if err != nil {
+			fmt.Printf("match: %s failed to read node_name file: %s", match, err.Error())
+			continue
+		}
+		FCHostInfo.NodeName = strings.TrimSpace(string(data))
+		FCHostsInfo = append(FCHostsInfo, FCHostInfo)
+	}
+
+	if len(FCHostsInfo) == 0 {
+		return []FCHBAInfo{}, err
+	}
+	return FCHostsInfo, nil
+}
+
+// GetInitiators returns a list of initiators on the local system.
+func (nvme *NVMeFC) GetInitiators(filename string) ([]string, error) {
+	return nvme.getInitiators(filename)
+}
+
+func (nvme *NVMeFC) getInitiators(filename string) ([]string, error) {
+
+	// a slice of filename, which might exist and define the nvme initiators
+	initiatorConfig := []string{}
+	nqns := []string{}
+
+	if filename == "" {
+		// add default filename(s) here
+		// /etc/nvme/hostnqn is the proper file for CentOS, RedHat, Sles, Ubuntu
+		if nvme.getChrootDirectory() != "/" {
+			initiatorConfig = append(initiatorConfig, nvme.getChrootDirectory()+"/"+DefaultInitiatorNameFile)
+		} else {
+			initiatorConfig = append(initiatorConfig, DefaultInitiatorNameFile)
+		}
+	} else {
+		initiatorConfig = append(initiatorConfig, filename)
+	}
+
+	var err error
+	// for each initiatior config file
+	for _, init := range initiatorConfig {
+		// make sure the file exists
+		_, err = os.Stat(init)
+		if err != nil {
+			continue
+		}
+
+		// get the contents of the initiator config file
+		// TODO: check if sys call is available for cat command
+		cmd := exec.Command("cat", init)
+
+		out, err := cmd.Output()
+		if err != nil {
+			fmt.Printf("Error gathering initiator names: %v", err)
+		}
+		lines := strings.Split(string(out), "\n")
+
+		for _, line := range lines {
+
+			if line != "" {
+				nqns = append(nqns, line)
+			}
+		}
+	}
+
+	if len(nqns) == 0 {
+		return nqns, err
+	}
+
+	return nqns, nil
+}
+
+// DiscoverNVMeTargets - runs NVMe/FC discovery and returns a list of targets.
+func (nvme *NVMeFC) DiscoverNVMeTargets(targetAddress string, login bool) ([]NVMeTarget, error) {
```

Comment:
Why do we need two functions each? DiscoverNVMeTargets and discoverNVMeTargets, GetInitiators() and getInitiators(), etc?